### DISCUSSION TITLE
Unexpected behaviour of BackgroundTask in AWS Lambda

### DISCUSSION THREAD

**Discussion by aniljaiswal** [fastapi/fastapi#7714] [Questions] [answered]
### First check

* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

I'm running FastAPI on AWS Lambda using [Magnum](https://github.com/erm/magnum) ASGI adapter like below where `handler` is the entry point for API endpoints specified in `serverless.yml`.

```
from fastapi import FastAPI
from mangum import Mangum

app = FastAPI()

handler = Mangum(app)
```

I have recently implemented a BackgroundTask on one of the route endpoints as documented in FastAPI which has to generates a CSV file and upload to S3 and then send an email to the user. The setup works fine locally using `uvicorn` (bypassing Magnum) and returns a quick HTTP response while the background task takes its own sweet time and does its work. 

However, when deployed on AWS Lambda, the request doesn't complete until the background task is also completed in the same request which increases the response latency and defeats the whole purpose of using a BackgroundTask.

What could be going wrong here?

### Additional context

Python: 3.8
FastAPI: 0.54

**Comment by aniljaiswal** [comment]
@tiangolo Any updates on this?

**Comment by phy25** [comment]
Reference: https://stackoverflow.com/questions/44668605/aws-lambda-to-run-in-background-even-after-sending-response-to-api-gateway

**Comment by aniljaiswal** [comment]
Thanks, @phy25 for the link. It looks like all synchronous invocations must complete fully before sending a response. I have made plans to use Step Functions to execute my workflow, just wanted to confirm if this is not limited by FastAPI in any way before moving ahead with the implementation.

**Comment by tiangolo** [comment]
Thanks for the help here @phy25 ! :clap: :bow: 

Thanks for reporting back and closing the issue @anil-grexit  :+1: 

Yeah, as I understand those function services like lambda start the app only for the request and kill it right after.

**Comment by abd1rahmane** [comment]
@tiangolo Google Cloud Function supports "Async **tasks**" .

**Comment by renkasiyas** [comment]
@aniljaiswal How did you solve this? I have the exact same setup.

**Comment by aniljaiswal** [comment, marked as answer]
@renkasiyas The way lambda works is that once the request is completed, the lambda returns a response and is ready to handle other requests. So any background tasks you queue in async fashion are lost. I am using SQS to push those background tasks and they are handled via a different lambda function.

**Comment by mfulgo** [comment]
It should now be possible to implement something that does what you want - run background tasks after the response is returned - by [creating a Lambda Extension](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-extensions-api.html).

The [LambdaPunch](https://github.com/rails-lambda/lambda_punch) gem does this for Ruby, and [DataDog's Lambda Extension](https://github.com/DataDog/datadog-lambda-extension) does similar things for its metrics collection. There's also a [python example](https://github.com/aws-samples/aws-lambda-extensions/blob/main/python-example-logs-api-extension/README.md)
